panoply/datasource.py

```diff
@@ -18,13 +21,58 @@ def log(self, *msgs):
         if "logger" in self.options:
             self.options["logger"](msgs)
         else:
-            print msgs
+            print(msgs)
 
-    def progress(self, loaded, total, msg):
+    def progress(self, loaded, total, msg=''):
         self.fire("progress", {
             "loaded": loaded,
             "total": total,
             "msg": msg
         })
 
+# a decorator used to invalidate the access_token for oauth based data sources
+# this should be used on every method in the data source that fetches data
+# from the server (controlled by this oauth), and that needs to be invalidated
+def invalidate_token(refresh_url, callback=None,
+                     keys=['access_token', 'refresh_token']):
+    def _invalidate_token(f):
+        def wrapper(*args):
+            self = args[0]
+            try:
+                return f(*args)
+            except Exception, e:
+                try:
+                    self.log("Reinvalidating access_token...")
+                    self.source[keys[0]] = None
```

Comment:
what if the refresh url is wrong, empty or a dictionary instead of a string?
The SDK is meant to be used how we defined it. If someone misuses it, he will probably get an error...